tests/test_diag_matrix.py

```diff
@@ -0,0 +1,630 @@
+        ref = AA * blk_diag.dense()  # zero out off block elem
+
+        # Test default unit weight scaling
+        x = d1.lr_scale(blk_diag.partition)
+        np.testing.assert_allclose(x.dense(), ref, atol=utest_tolerance(d1.dtype))
+
+        # Test with different weights
+        w = np.arange(d1.count, dtype=d1.dtype)
+        x = d1.lr_scale(blk_diag.partition, weights=w)
+        ref = (w * AA) * blk_diag.dense()  # zero out off block elem
+        np.testing.assert_allclose(x.dense(), ref, atol=utest_tolerance(d1.dtype))
+
+
+def test_bad_partition():
+    """
+    Test incorrect partitions raise an error.
+    """
+    d = DiagMatrix(np.empty(8))
+
+    partition = [(9, 9)]
+    with pytest.raises(RuntimeError, match=r".*count.*does not match.*"):
+        _ = d.lr_scale(partition)
+
+    partition = [(4, 4), (3, 4)]
+    with pytest.raises(RuntimeError, match=r".*was not square.*"):
+        _ = d.lr_scale(partition)
+
+
+def test_as_blk_diag(matrix_size, blk_diag):
+    """
+    Test conversion from `DiagMatrix` to `BlkDiagMatrix`.
+
+    Note this relies on `BlkDiagMatrix.dense`.
+    """
+
+    # Construct via Numpy.
+    d_np = np.random.randn(matrix_size).astype(blk_diag.dtype)
+    A = np.diag(d_np)
+
+    # Create DiagMatrix then convert to BlkDiagMatrix
+    d = DiagMatrix(d_np)
+    B = d.as_blk_diag(blk_diag.partition)
+
+    # Compare the dense B with the dense A.
+    np.testing.assert_equal(B.dense(), A)
+
+
+def test_bad_as_blk_diag(matrix_size, blk_diag):
+    """
+    Test unimplemented conversion raises.
+    """
+    with pytest.raises(RuntimeError, match=r".*only implemented for singletons.*"):
+        # Construct via Numpy.
+        d_np = np.empty((2, matrix_size), dtype=blk_diag.dtype)
+
+        # Create DiagMatrix then convert to BlkDiagMatrix
+        d = DiagMatrix(d_np)
+        _ = d.as_blk_diag(blk_diag.partition)
+
+
+def test_eigs(diag_matrix_fixture):
+    """
+    Test the `eigvals` util method.
+    """
+    d, _, d_np = diag_matrix_fixture
+
+    np.testing.assert_equal(d.eigvals(), d_np[0])
+
+
+def test_eye(matrix_size, dtype):
+    """
+    Test helper for identity matrix.
+    Same as `ones` for `DiagMatrix`.
+    """
+    d = DiagMatrix.eye(matrix_size, dtype=dtype)
+
+    np.testing.assert_equal(d, np.ones(matrix_size, dtype=dtype))
+
+
+def test_apply(diag_matrix_fixture):
+    """
+    Test the `apply` method against numpy.
+    """
+    d1, _, d_np = diag_matrix_fixture
+
+    x = d1.apply(d_np)
+
+    np.testing.assert_allclose(x, d_np[0][None, :] * d_np)
+
+
+def test_rapply(diag_matrix_fixture):
+    """
+    Test the `rapply` method against numpy.
+    """
+
+    d1, _, d_np = diag_matrix_fixture
+
+    x = d1.rapply(d_np)
+
+    np.testing.assert_allclose(x, (d_np * d_np[0]))
+
+
+def test_solve(diag_matrix_fixture):
+    """
+    Test `solve` output is a valid solution.
+    """
+    a, _, d_np = diag_matrix_fixture
+
+    b = np.arange(a.count, dtype=a.dtype)
```

Comment:
Should we test with a `DiagMatrix` instance as well? The docstring for `solve()` labels `b` as such.

Reply:
i think the docstring was just wrong 😇 .

Reply:
Updated docstring param `b` as np array.